packages/runtime/runtime-utils/src/summarizerNode/summarizerNodeWithGc.ts

```diff
@@ -270,19 +271,59 @@ export class SummarizerNodeWithGC extends SummarizerNode implements IRootSummari
+             * routes of all its children and adds it to their snapshot tree.
+             * All the other nodes gets the GC data and used routes from their snapshot tree and updates their state.
+             * They get the GC data and used routes of their children and add it to their snapshot tree and so on.
+             */
+            let gcDetails: IGarbageCollectionDetailsBase = { gcData: { gcNodes: {} } };
+            const gcSnapshotTree = snapshotTree.trees[gcTreeKey];
+            if (gcSnapshotTree !== undefined) {
+                // This is the root summarizer node. It reads GC data from the root and processes it as explained above.
+                const gcSnapshotData = await getGCDataFromSnapshot(gcSnapshotTree, readAndParseBlob);
 
                 // Possible re-entrancy. If we have already seen a summary later than this one, ignore it.
                 if (this.referenceSequenceNumber >= referenceSequenceNumber) {
                     return;
                 }
 
-                this.referenceUsedRoutes = gcDetails.usedRoutes;
+                const gcNodes: { [id: string]: string[]; } = {};
+                for (const [nodeId, nodeData] of Object.entries(gcSnapshotData.gcState.gcNodes)) {
+                    gcNodes[nodeId] = Array.from(nodeData.outboundRoutes);
+                }
+                // Run GC on the nodes in the snapshot to get the used routes for each node in the container.
+                const usedRoutes = runGarbageCollection(gcNodes, ["/"]).referencedNodeIds;
+                gcDetails = { gcData: { gcNodes }, usedRoutes };
+            } else {
+                // Non root summarizer nodes. They get their GC data and used routes from their snapshot tree.
+                const gcDetailsBlob = snapshotTree.blobs[gcTreeKey];
+                gcDetails = JSON.parse(gcDetailsBlob) as IGarbageCollectionDetailsBase;
+
+                // Possible re-entrancy. If we have already seen a summary later than this one, ignore it.
+                if (this.referenceSequenceNumber >= referenceSequenceNumber) {
+                    return;
+                }
             }
+
+            // Update the GC data and reference used routes for this node.
+            this.gcData = gcDetails.gcData !== undefined ? cloneGCData(gcDetails.gcData) : undefined;
+            this.referenceUsedRoutes = gcDetails.usedRoutes !== undefined ? Array.from(gcDetails.usedRoutes) : undefined;
+
+            // Generate the GC data and used routes of children GC nodes and add it to their snapshot tree.
+            const baseGCDetailsMap = unpackChildNodesGCDetails(gcDetails);
+            const { childrenTree } = parseSummaryForSubtrees(snapshotTree);
+            baseGCDetailsMap.forEach((childGCDetails: IGarbageCollectionDetailsBase, childId: string) => {
+                childrenTree.trees[childId].blobs[gcTreeKey] = JSON.stringify(childGCDetails);
```

Comment:
I'm thinking that we already have access to all our children, why do we need to put this information into the tree itself? Why don't we just set `this.gcData` of the children, and that way the difference between root data, node data, and no data is clear.

Reply:
All the summarizer nodes are SummarizerNodesWithGC anyway, is it expensive to change the contract?

Reply:
I am not sure I understand what you mean by root data, node data and no data.

This solution is even more hacky that the current one (which is sort of a hack too). Directly setting the gcData of the children here is not a good idea because of the following reasons:
- The child node's state must be taken into account before setting the data. For example, the reference sequence number of the children has to be checked similar to what is happening here.
- Child node's loadBaseGCDetails must be called before setting the GC data or it will get overriden when GC runs.
- It adds yet another flow to the already complicated machinery.

The current solution, while complex, has the advantage that its consistent with the existing flows and design. It also lets each node decide what it wants to do with this data which is important.

Reply:
I'm ok with the current solution then :)